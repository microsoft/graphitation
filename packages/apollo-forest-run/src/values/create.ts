import type {
  FieldInfo,
  OperationDescriptor,
  PossibleSelections,
  TypeName,
} from "../descriptor/types";
import type {
  ComplexScalarValue,
  CompositeListAggregate,
  CompositeListChunk,
  CompositeListValue,
  CompositeNullAggregate,
  CompositeNullChunk,
  CompositeNullValue,
  CompositeUndefinedAggregate,
  CompositeUndefinedChunk,
  CompositeUndefinedValue,
  CompositeValue,
  CompositeValueChunk,
  FormattedError,
  GraphValue,
  LeafErrorValue,
  LeafListValue,
  ObjectAggregate,
  ObjectChunk,
  ObjectValue,
  SourceCompositeList,
  SourceCompositeValue,
  SourceCustomScalar,
  SourceLeafList,
  SourceNull,
  SourceObject,
} from "./types";
import { ValueKind } from "./types";
import * as Predicates from "./predicates";
import { assert, assertNever } from "../jsutils/assert";
import { resolveSelection } from "../descriptor/resolvedSelection";

export type ChunkAccumulator = {
  obj: ObjectChunk[] | undefined;
  list: CompositeListChunk[] | undefined;
  nll: CompositeNullChunk[] | undefined;
  undef: CompositeUndefinedChunk[] | undefined;
  err: LeafErrorValue | undefined;
};

export function shouldAggregateChunks(
  value: GraphValue,
): value is CompositeValue | LeafErrorValue {
  return (
    Predicates.isCompositeValue(value) || Predicates.isLeafErrorValue(value)
  );
}

export function createChunkAccumulator(): ChunkAccumulator {
  return {
    obj: undefined,
    list: undefined,
    nll: undefined,
    undef: undefined,
    err: undefined,
  };
}
export function accumulateChunks(
  accumulator: ChunkAccumulator,
  value: CompositeValue | LeafErrorValue,
): void {
  // Taking into account the following case
  //   (first "foo" value is null because of error bubbling, but the second is actually OK)
  // ```graphql
  // {
  //     a: foo { bar }
  //     b: foo { baz }
  // }
  // ```
  //
  // ```js
  // const data = {
  //     "a": null,
  //     "b": [{ baz: "baz" }]
  // }
  // ```
  if (value === null) {
    return;
  }
  switch (value.kind) {
    case ValueKind.Object:
      return accumulateObjectChunks(accumulator, value);

    case ValueKind.CompositeList:
      return accumulateListChunks(accumulator, value);

    case ValueKind.CompositeNull:
      return accumulateNullChunks(accumulator, value);

    case ValueKind.CompositeUndefined:
      return accumulateUndefinedChunks(accumulator, value);

    case ValueKind.LeafError: {
      accumulator.err = value;
      return;
    }
    default:
      assertNever(value);
  }
}

export function createValue(
  accumulator: ChunkAccumulator,
): CompositeValue | LeafErrorValue | undefined {
  const { err, list, obj, nll, undef } = accumulator;

  if (obj) {
    assert(!list && !err);
    return obj.length === 1 && !nll && !undef
      ? obj[0]
      : createObjectAggregate(obj, nll, undef);
  }
  if (list) {
    assert(!obj && !err);

    return list.length === 1 && !nll
      ? list[0]
      : createCompositeListAggregate(list, nll);
  }
  if (nll) {
    assert(!err);
    return nll.length === 1 ? nll[0] : createCompositeNullAggregate(nll);
  }
  if (undef) {
    assert(!err);
    return undef.length === 1
      ? undef[0]
      : createCompositeUndefinedAggregate(undef);
  }
  if (err) {
    return err;
  }
  return undefined;
}

function accumulateObjectChunks(
  accumulator: ChunkAccumulator,
  value: ObjectValue,
) {
  accumulator.obj ??= [];

  if (!Predicates.isAggregate(value)) {
    accumulator.obj.push(value);
    return;
  }
  accumulator.obj.push(...value.chunks);

  if (value.nullChunks?.length) {
    accumulator.nll ??= [];
    accumulator.nll.push(...value.nullChunks);
  }
}

function accumulateListChunks(
  accumulator: ChunkAccumulator,
  value: CompositeListValue,
) {
  accumulator.list ??= [];

  if (!Predicates.isAggregate(value)) {
    accumulator.list.push(value);
    return;
  }
  accumulator.list.push(...value.chunks);

  if (value.nullChunks?.length) {
    accumulator.nll ??= [];
    accumulator.nll.push(...value.nullChunks);
  }
}

function accumulateNullChunks(
  accumulator: ChunkAccumulator,
  value: CompositeNullValue,
) {
  accumulator.nll ??= [];

  if (Predicates.isAggregate(value)) {
    accumulator.nll.push(...value.chunks);
  } else {
    accumulator.nll.push(value);
  }
}

function accumulateUndefinedChunks(
  accumulator: ChunkAccumulator,
  value: CompositeUndefinedValue,
) {
  accumulator.undef ??= [];

  if (Predicates.isAggregate(value)) {
    accumulator.undef.push(...value.chunks);
  } else {
    accumulator.undef.push(value);
  }
}

export function createLeafError(error: FormattedError): LeafErrorValue {
  return {
    kind: ValueKind.LeafError,
    data: null as SourceNull,
    error,
  };
}

export function createLeafList(source: SourceLeafList): LeafListValue {
  return {
    kind: ValueKind.LeafList,
    data: source,
  };
}

export function createObjectChunk(
  operation: OperationDescriptor,
  possibleSelections: PossibleSelections,
  source: SourceObject,
  key: string | false,
  missingFields: Set<FieldInfo> | null = null,
): ObjectChunk {
  let typeName = source.__typename;
  if (!typeName && key !== false && key === operation.rootNodeKey) {
    typeName = operation.rootType;
  }
  return {
    kind: ValueKind.Object,
    isAggregate: false,
    operation,
    data: source,
    possibleSelections,
    // TODO: resolveSelection should be passed here instead
    selection: resolveSelection(
      operation,
      possibleSelections,
      typeName || null,
    ),
    fieldChunks: new Map(),
    type: typeName ?? false,
    key,
    missingFields,
    partialFields: null,
    hasNestedReadPolicies: false,
  };
}

export function createCompositeListChunk(
  operation: OperationDescriptor,
  possibleSelections: PossibleSelections,
  source: SourceCompositeList,
): CompositeListChunk {
  return {
    kind: ValueKind.CompositeList,
    isAggregate: false,
    operation,
    data: source,
    possibleSelections,
    itemChunks: new Array(source.length),
    hasNestedReadPolicies: false,
    missingItems: null,
    partialItems: null,
  };
}

export function createCompositeNullChunk(
  operation: OperationDescriptor,
  possibleSelections: PossibleSelections,
): CompositeNullChunk {
  return {
    kind: ValueKind.CompositeNull,
    isAggregate: false,
    data: null as SourceNull,
    operation,
    possibleSelections,
  };
}

export function createCompositeUndefinedChunk(
  operation: OperationDescriptor,
  possibleSelections: PossibleSelections,
  deleted = false,
): CompositeUndefinedChunk {
  return {
    kind: ValueKind.CompositeUndefined,
    isAggregate: false,
    data: undefined,
    deleted,
    operation,
    possibleSelections,
  };
}

export function createCompositeValueChunk(
  operation: OperationDescriptor,
  possibleSelections: PossibleSelections,
  value: SourceCompositeValue,
  key?: string | false,
): CompositeValueChunk {
  if (value === null) {
    return createCompositeNullChunk(operation, possibleSelections);
  }
  if (value === undefined) {
    return createCompositeUndefinedChunk(operation, possibleSelections);
  }
  if (Array.isArray(value)) {
    return createCompositeListChunk(operation, possibleSelections, value);
  }
  if (key === undefined) {
    key = operation.env.objectKey?.(value) ?? false;
  }
  return createObjectChunk(operation, possibleSelections, value, key);
}

export function createObjectAggregate(
  chunks: ObjectChunk[],
  nullChunks?: CompositeNullChunk[],
  undefinedChunks?: CompositeUndefinedChunk[],
): ObjectAggregate {
  if (!chunks.length) {
    throw new Error("Object chunks are empty");
  }
  const chunk = chunks[0];
  return {
    kind: ValueKind.Object,
    isAggregate: true,
    data: chunk.data,
    key: chunk.key,
    type:
      chunk.type ||
      (chunks.find((chunk) => chunk.type !== false)?.type ?? false),
    chunks,
    nullChunks,
    undefinedChunks,
  };
}

export function createCompositeListAggregate(
  chunks: CompositeListChunk[],
  nullChunks?: CompositeNullChunk[],
): CompositeListAggregate {
  if (!chunks.length) {
    throw new Error("List chunks are empty");
  }
  return {
    kind: ValueKind.CompositeList,
    isAggregate: true,
    data: chunks[0].data,
    chunks,
    nullChunks,
  };
}

export function createCompositeNullAggregate(
  chunks: CompositeNullChunk[],
): CompositeNullAggregate {
  if (!chunks.length) {
    throw new Error("List chunks are empty");
  }
  return {
    kind: ValueKind.CompositeNull,
    isAggregate: true,
    data: null as SourceNull,
    chunks,
  };
}

export function createCompositeUndefinedAggregate(
  chunks: CompositeUndefinedChunk[],
): CompositeUndefinedAggregate {
  if (!chunks.length) {
    throw new Error("List chunks are empty");
  }
  return {
    kind: ValueKind.CompositeUndefined,
    isAggregate: true,
    data: undefined,
    deleted: chunks[0].deleted, // assuming _all_ deleted chunks are marked as deleted, so relying on a single chunk
    chunks,
  };
}

export function createComplexScalarValue(
  source: SourceCustomScalar,
): ComplexScalarValue {
  return {
    kind: ValueKind.ComplexScalar,
    data: source,
  };
}

export const createSourceObject = (
  typename?: TypeName | object,
): SourceObject =>
  typeof typename === "object"
    ? (typename as SourceObject)
    : ({
        __typename: typename,
      } as SourceObject);
