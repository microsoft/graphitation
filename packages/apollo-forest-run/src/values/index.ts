export * from "./create";
export * from "./delete";
export * from "./draft";
export * from "./iterator";
export * from "./traverse";
export * from "./predicates";
export * from "./resolve";
