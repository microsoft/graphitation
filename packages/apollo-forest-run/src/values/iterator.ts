import { CompositeListValue, ObjectValue, ValueKind } from "./types";
import { aggregateListItemValue } from "./resolve";
import { assertNever } from "../jsutils/assert";

export type IterableListValue = {
  readonly list: CompositeListValue;
  readonly length: number;
  [Symbol.iterator](): Iterator<ObjectValue | null | IterableListValue>;
};

type NextResult = {
  done: boolean;
  value: ObjectValue | IterableListValue | null;
};

export function toIterableValue(list: CompositeListValue): IterableListValue {
  return { list, length: list.data.length, [Symbol.iterator]: iterator };
}

function iterator(
  this: IterableListValue,
): Iterator<ObjectValue | IterableListValue | null> {
  const list = this.list;
  const len = list.data.length;
  const next: NextResult = { done: false, value: null };
  let i = 0;
  return {
    next() {
      if (i >= len) {
        next.done = true;
        return next;
      }
      const item = aggregateListItemValue(list, i++);
      if (item.kind === ValueKind.Object) {
        next.value = item;
        return next;
      }
      if (item.kind === ValueKind.CompositeList) {
        next.value = toIterableValue(item);
        return next;
      }
      if (item.kind === ValueKind.CompositeNull) {
        next.value = null;
        return next;
      }
      if (item.kind === ValueKind.CompositeUndefined) {
        const itemChunk = item.isAggregate ? item.chunks[0] : item;
        throw new Error(
          `Missing list item ${i - 1} in ${JSON.stringify(list)}\n` +
            `  operation: ${itemChunk.operation.definition.name?.value}`,
        );
      }
      assertNever(item);
    },
  };
}
