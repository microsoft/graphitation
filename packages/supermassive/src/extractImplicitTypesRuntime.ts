// Given a GraphQL document with type definitions, extract all types that can be defined implicitly.
import {
  DocumentNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  getDirectiveValues,
  ObjectTypeDefinitionNode,
  GraphQLDeprecatedDirective,
  NamedTypeNode,
  GraphQLEnumType,
  GraphQLEnumValueConfigMap,
  GraphQLScalarType,
  InputValueDefinitionNode,
  Kind,
  EnumTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  GraphQLInputObjectType,
  GraphQLInputFieldConfigMap,
  GraphQLInputType,
  valueFromAST,
  TypeNode,
  GraphQLNonNull,
  GraphQLList,
} from "graphql";
import { Maybe } from "./jsutils/Maybe";
import { Resolvers, UnionTypeResolver, InterfaceTypeResolver } from "./types";

export function extractImplicitTypes<TSource = unknown, TContext = unknown>(
  document: DocumentNode,
  getTypeByName: (name: string) => GraphQLInputType,
): Resolvers<TSource, TContext> {
  const result: Resolvers<TSource, TContext> = Object.create(null);
  const implementedBy: Record<string, Array<string>> = {};
  for (const astNode of document.definitions) {
    if (astNode.kind === Kind.SCALAR_TYPE_DEFINITION) {
      const name = astNode.name.value;
      result[name] = new GraphQLScalarType({
        name,
        description: astNode.description?.value,
      });
    } else if (astNode.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION) {
      result[astNode.name.value] = makeInputObject(astNode, getTypeByName);
    } else if (astNode.kind === Kind.ENUM_TYPE_DEFINITION) {
      result[astNode.name.value] = makeEnum(astNode);
    } else if (astNode.kind === Kind.INTERFACE_TYPE_DEFINITION) {
      if (!implementedBy[astNode.name.value]) {
        implementedBy[astNode.name.value] = [];
      }
      result[astNode.name.value] = {
        __resolveType: undefined,
        __implementedBy: implementedBy[astNode.name.value],
      } as InterfaceTypeResolver;
    } else if (astNode.kind === Kind.UNION_TYPE_DEFINITION) {
      const types = astNode.types?.map((typeNode) => {
        return typeNode.name.value;
      });

      result[astNode.name.value] = {
        __resolveType: undefined,
        __types: types || [],
      } as UnionTypeResolver;
    } else if (astNode.kind === Kind.OBJECT_TYPE_DEFINITION) {
      astNode.interfaces?.forEach((node: NamedTypeNode) => {
        if (!implementedBy[node.name.value]) {
          implementedBy[node.name.value] = [];
        }
        implementedBy[node.name.value].push(
          (astNode as ObjectTypeDefinitionNode).name.value,
        );
      });

      result[astNode.name.value] = {};
    }
  }
  return result;
}

function makeEnum(astNode: EnumTypeDefinitionNode) {
  const enumValueMap: GraphQLEnumValueConfigMap = Object.create(null);
  for (const value of astNode.values || []) {
    enumValueMap[value.name.value] = {
      description: value.description?.value,
      deprecationReason: getDeprecationReason(value),
      astNode: value,
    };
  }

  return new GraphQLEnumType({
    name: astNode.name.value,
    description: astNode.description?.value,
    values: enumValueMap,
  });
}

function makeInputObject(
  astNode: InputObjectTypeDefinitionNode,
  getTypeByName: (name: string) => GraphQLInputType,
) {
  const name = astNode.name.value;

  return new GraphQLInputObjectType({
    name,
    description: astNode.description?.value,
    fields: () => buildInputFieldMap(astNode.fields || [], getTypeByName),
    astNode,
  });
}

function buildInputFieldMap(
  fieldGroup: ReadonlyArray<InputValueDefinitionNode>,
  getTypeByName: (name: string) => GraphQLInputType,
): GraphQLInputFieldConfigMap {
  const inputFieldMap = Object.create(null);
  for (const field of fieldGroup) {
    // Note: While this could make assertions to get the correctly typed
    // value, that would throw immediately while type system validation
    // with validateSchema() will produce more actionable results.
    const type: GraphQLInputType = getWrappedType(field.type, getTypeByName);

    inputFieldMap[field.name.value] = {
      type,
      description: field.description?.value,
      defaultValue: valueFromAST(field.defaultValue, type),
      deprecationReason: getDeprecationReason(field),
      astNode: field,
    };
  }
  return inputFieldMap;
}

function getWrappedType(
  type: TypeNode,
  getTypeByName: (name: string) => GraphQLInputType,
): GraphQLInputType {
  if (type.kind === Kind.LIST_TYPE) {
    return new GraphQLList(getWrappedType(type.type, getTypeByName));
  } else if (type.kind === Kind.NON_NULL_TYPE) {
    return new GraphQLNonNull(getWrappedType(type.type, getTypeByName));
  } else {
    return getTypeByName(type.name.value);
  }
}

function getDeprecationReason(
  node:
    | EnumValueDefinitionNode
    | FieldDefinitionNode
    | InputValueDefinitionNode,
): Maybe<string> {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node) as
    | { reason: string }
    | undefined;
  return deprecated?.reason;
}
