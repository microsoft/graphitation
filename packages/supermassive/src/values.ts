import {
  GraphQLError,
  Kind,
  print,
  locatedError,
  DirectiveNode,
  FieldNode,
  VariableDefinitionNode,
} from "graphql";
import { inspect } from "./jsutils/inspect";
import { printPathArray } from "./jsutils/printPathArray";
import { ExecutionContext } from "./executeWithoutSchema";
import {
  DirectiveDefinitionTuple,
  FieldDefinition,
  getFieldArguments,
  getDirectiveName,
  getInputDefaultValue,
  getInputValueTypeReference,
  isDefined,
  isInputType,
  getDirectiveArguments,
} from "./schema/definition";
import { valueFromAST } from "./utilities/valueFromAST";
import { coerceInputValue } from "./utilities/coerceInputValue";
import {
  inspectTypeReference,
  isNonNullType,
  typeReferenceFromNode,
} from "./schema/reference";
import type { SchemaFragment } from "./types";

type CoercedVariableValues =
  | { errors: Array<GraphQLError>; coerced?: never }
  | { coerced: { [variable: string]: unknown }; errors?: never };

/**
 * Prepares an object map of variableValues of the correct type based on the
 * provided variable definitions and arbitrary input. If the input cannot be
 * parsed to match the variable definitions, a GraphQLError will be thrown.
 *
 * Note: The returned value is a plain Object with a prototype, since it is
 * exposed to user code. Care should be taken to not pull values from the
 * Object prototype.
 *
 * @internal
 */
export function getVariableValues(
  schemaFragment: SchemaFragment,
  varDefNodes: ReadonlyArray<VariableDefinitionNode>,
  inputs: { [variable: string]: unknown },
  options?: { maxErrors?: number },
): CoercedVariableValues {
  const errors: GraphQLError[] = [];
  const maxErrors = options?.maxErrors;
  try {
    const coerced = coerceVariableValues(
      schemaFragment,
      varDefNodes,
      inputs,
      (error) => {
        if (maxErrors != null && errors.length >= maxErrors) {
          throw locatedError(
            "Too many errors processing variables, error limit reached. Execution aborted.",
            [],
          );
        }
        errors.push(error);
      },
    );

    if (errors.length === 0) {
      return { coerced };
    }
  } catch (error) {
    errors.push(error as GraphQLError);
  }

  return { errors: errors };
}

function coerceVariableValues(
  schemaFragment: SchemaFragment,
  varDefNodes: ReadonlyArray<VariableDefinitionNode>,
  inputs: { [variable: string]: unknown },
  onError: (error: GraphQLError) => void,
): { [variable: string]: unknown } {
  const coercedValues: { [variable: string]: unknown } = {};
  for (const varDefNode of varDefNodes) {
    const varName = varDefNode.variable.name.value;
    const varTypeReference = typeReferenceFromNode(varDefNode.type);

    if (!isInputType(schemaFragment.definitions, varTypeReference)) {
      // Must use input types for variables. This should be caught during
      // validation, however is checked again here for safety.
      const varTypeStr = inspectTypeReference(varTypeReference);
      onError(
        locatedError(
          `Variable "$${varName}" expected value of type "${varTypeStr}" which cannot be used as an input type.`,
          [varDefNode.type],
        ),
      );
      continue;
    }

    if (!hasOwnProperty(inputs, varName)) {
      if (varDefNode.defaultValue) {
        coercedValues[varName] = valueFromAST(
          varDefNode.defaultValue,
          varTypeReference,
          schemaFragment,
        );
      } else if (isNonNullType(varTypeReference)) {
        const varTypeStr = inspectTypeReference(varTypeReference);
        onError(
          locatedError(
            `Variable "$${varName}" of required type "${varTypeStr}" was not provided.`,
            [varDefNode],
          ),
        );
      }
      continue;
    }

    const value = inputs[varName];
    if (value === null && isNonNullType(varTypeReference)) {
      const varTypeStr = inspectTypeReference(varTypeReference);
      onError(
        locatedError(
          `Variable "$${varName}" of non-null type "${varTypeStr}" must not be null.`,
          [varDefNode],
        ),
      );
      continue;
    }

    coercedValues[varName] = coerceInputValue(
      value,
      varTypeReference,
      schemaFragment,
      (path, invalidValue, error) => {
        let prefix =
          `Variable "$${varName}" got invalid value ` + inspect(invalidValue);
        if (path.length > 0) {
          prefix += ` at "${varName}${printPathArray(path)}"`;
        }
        onError(locatedError(prefix + "; " + error.message, [varDefNode]));
      },
    );
  }

  return coercedValues;
}

/**
 * Prepares an object map of argument values given a list of argument
 * definitions and list of argument AST nodes.
 *
 * Note: The returned value is a plain Object with a prototype, since it is
 * exposed to user code. Care should be taken to not pull values from the
 * Object prototype.
 *
 * @internal
 */
export function getArgumentValues(
  exeContext: ExecutionContext,
  def: FieldDefinition | DirectiveDefinitionTuple,
  node: FieldNode | DirectiveNode,
): { [argument: string]: unknown } {
  const definitions = exeContext.schemaFragment.definitions;
  const coercedValues: { [argument: string]: unknown } = {};
  const argumentDefs =
    node.kind === Kind.FIELD
      ? getFieldArguments(def as FieldDefinition)
      : getDirectiveArguments(def as DirectiveDefinitionTuple);
  if (!argumentDefs) {
    return coercedValues;
  }
  // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
  const argumentNodes = node.arguments ?? [];
  const argNodeMap = new Map(argumentNodes.map((arg) => [arg.name.value, arg]));

  for (const [name, argumentDef] of Object.entries(argumentDefs)) {
    const argumentNode = argNodeMap.get(name);
    const argumentTypeRef = getInputValueTypeReference(argumentDef);
    const defaultValue = getInputDefaultValue(argumentDef);

    if (argumentNode == null) {
      if (defaultValue !== undefined) {
        coercedValues[name] = defaultValue;
      } else if (isNonNullType(argumentTypeRef)) {
        const type = inspectTypeReference(argumentTypeRef);
        throw locatedError(
          `Argument "${name}" of required type "${type}" was not provided.`,
          [node],
        );
      }
      continue;
    }

    if (!isDefined(definitions, argumentTypeRef)) {
      throw locatedError(
        `Could not find type for argument ${name} in ${node.kind} ${node.name.value}`,
        [argumentNode],
      );
    }

    // if (!schemaTypes.isInputType(argumentTypeRef)) {
    //   const type = schemaTypes.printTypeRef(argumentTypeRef);
    //   throw locatedError(
    //     `Argument "$${name}" expected value of type "${type}" which cannot be used as an input type.`,
    //     [argumentNode],
    //   );
    // }

    const valueNode = argumentNode.value;
    let isNull = valueNode.kind === Kind.NULL;

    if (valueNode.kind === Kind.VARIABLE) {
      const variableName = valueNode.name.value;
      if (
        exeContext.variableValues == null ||
        !hasOwnProperty(exeContext.variableValues, variableName)
      ) {
        if (defaultValue !== undefined) {
          coercedValues[name] = defaultValue;
        } else if (isNonNullType(argumentTypeRef)) {
          const type = inspectTypeReference(argumentTypeRef);
          throw locatedError(
            `Argument "${name}" of required type "${type}" ` +
              `was provided the variable "$${variableName}" which was not provided a runtime value.`,
            [valueNode],
          );
        }
        continue;
      }
      isNull = exeContext.variableValues[variableName] == null;
    }

    if (isNull && isNonNullType(argumentTypeRef)) {
      const type = inspectTypeReference(argumentTypeRef);
      throw locatedError(
        `Argument "${name}" of non-null type "${type}" must not be null."`,
        [valueNode],
      );
    }

    const coercedValue = valueFromAST(
      valueNode,
      argumentTypeRef,
      exeContext.schemaFragment,
      exeContext.variableValues,
    );
    if (coercedValue === undefined) {
      // Note: ValuesOfCorrectTypeRule validation should catch this before
      // execution. This is a runtime check to ensure execution does not
      // continue with an invalid argument value.
      throw locatedError(
        `Argument "${name}" has invalid value ${print(valueNode)}.`,
        [valueNode],
      );
    }
    coercedValues[name] = coercedValue;
  }

  return coercedValues;
}

/**
 * Prepares an object map of argument values given a directive definition
 * and a AST node which may contain directives. Optionally also accepts a map
 * of variable values.
 *
 * If the directive does not exist on the node, returns undefined.
 *
 * Note: The returned value is a plain Object with a prototype, since it is
 * exposed to user code. Care should be taken to not pull values from the
 * Object prototype.
 */
export function getDirectiveValues(
  exeContext: ExecutionContext,
  directiveDef: DirectiveDefinitionTuple,
  node: { directives?: ReadonlyArray<DirectiveNode> },
): undefined | { [argument: string]: unknown } {
  const name = getDirectiveName(directiveDef);

  // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
  const directiveNode = node.directives?.find(
    (directive) => directive.name.value === name,
  );

  if (directiveNode) {
    return getArgumentValues(exeContext, directiveDef, directiveNode);
  }
}

function hasOwnProperty(obj: unknown, prop: string): boolean {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}
