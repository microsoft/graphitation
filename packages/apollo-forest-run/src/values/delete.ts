import { assertNever } from "../jsutils/assert";
import { FieldInfo } from "../descriptor/types";
import {
  CompositeListChunk,
  ObjectChunk,
  GraphChunkReference,
  CompositeValueChunkReference,
} from "./types";
import {
  isDeletedValue,
  isParentListRef,
  isParentObjectRef,
  isRootRef,
} from "./predicates";
import { createCompositeUndefinedChunk } from "./create";
import { resolveFieldChunk } from "./resolve";
import { TraverseEnv } from "./traverse";

export function markAsPartial(
  env: TraverseEnv,
  ref: GraphChunkReference | null,
) {
  if (!ref || isRootRef(ref)) {
    return;
  }
  if (isParentObjectRef(ref)) {
    const { parent, field } = ref;
    if (parent.partialFields?.has(field)) {
      return;
    }
    parent.partialFields ??= new Set();
    parent.partialFields.add(field);
    const parentRef = env.findParent(parent);
    markAsPartial(env, parentRef || null);
    return;
  }
  if (isParentListRef(ref)) {
    const { parent, index } = ref;
    if (parent.partialItems?.has(index)) {
      return;
    }
    parent.partialItems ??= new Set();
    parent.partialItems.add(index);
    const parentRef = env.findParent(parent);
    markAsPartial(env, parentRef || null);
    return;
  }
  assertNever(ref);
}

export function deleteField(
  env: TraverseEnv,
  chunk: ObjectChunk,
  fieldInfo: FieldInfo,
): boolean {
  const chunkRef = env.findParent(chunk);
  const hasField = chunk.selection.fields
    .get(fieldInfo.name)
    ?.some((field) => field === fieldInfo);

  if (!hasField || chunk.missingFields?.has(fieldInfo)) {
    return false;
  }
  // ApolloCompat: cache.modify allows to "delete" field values
  // TODO: remove DELETE support for cache modify and the whole "missing fields" / "missing items" notion
  //   instead we should always have all fields in place, but some of them should be marked as "stale" to
  //   indicate that they shouldn't be used for diffing and should be re-fetched from the server when possible
  chunk.missingFields ??= new Set();
  chunk.missingFields.add(fieldInfo);

  if (fieldInfo.selection) {
    const parentInfo: CompositeValueChunkReference = {
      value: createCompositeUndefinedChunk(
        chunk.operation,
        fieldInfo.selection,
        true,
      ),
      parent: chunk,
      field: fieldInfo,
    };
    chunk.fieldChunks.set(fieldInfo.dataKey, parentInfo);
  }
  markAsPartial(env, chunkRef);

  // Note: not mutating the source value on purpose, because it could be used by product code,
  //   instead just marking this value as "missing"
  // chunk.source[fieldInfo.dataKey] = undefined;
  return true;
}

export function deleteListItem(
  env: TraverseEnv,
  chunk: CompositeListChunk,
  index: number,
): boolean {
  if (index >= chunk.data.length || chunk.missingItems?.has(index)) {
    return false;
  }
  const chunkRef = env.findParent(chunk);

  chunk.missingItems ??= new Set();
  chunk.missingItems.add(index);
  chunk.itemChunks[index] = {
    value: createCompositeUndefinedChunk(
      chunk.operation,
      chunk.possibleSelections,
      true,
    ),
    parent: chunk,
    index,
  };
  markAsPartial(env, chunkRef);

  // Note: not mutating the source value on purpose, because it could be used by product code,
  //   instead just marking this value as "missing"
  // chunk.source[index] = undefined;
  return true;
}

export function hasDeletedField(chunk: ObjectChunk): boolean {
  if (!chunk.missingFields?.size) {
    return false;
  }
  for (const field of chunk.missingFields) {
    const fieldValue = resolveFieldChunk(chunk, field);
    if (isDeletedValue(fieldValue)) {
      return true;
    }
  }
  return false;
}
