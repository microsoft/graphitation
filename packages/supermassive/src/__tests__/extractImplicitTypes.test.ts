import * as fs from "fs";
import * as path from "path";
import { isInputType, parse } from "graphql";
import { extractImplicitTypes } from "../extractImplicitTypesRuntime";
import { specifiedScalars } from "../values";
import { Resolvers } from "..";

describe(extractImplicitTypes, () => {
  it("benchmark schema extract", () => {
    expect.assertions(1);
    const typeDefs = fs.readFileSync(
      path.join(
        __dirname,
        "../../../../examples/supermassive-benchmarks/src/swapi-schema/schema.graphql",
      ),
      {
        encoding: "utf-8",
      },
    );
    let resolvers: Resolvers = {};
    const getTypeByName = (name: string) => {
      const type = specifiedScalars[name] || resolvers[name];
      if (isInputType(type)) {
        return type;
      } else {
        throw new Error("Invalid type");
      }
    };
    resolvers = extractImplicitTypes(parse(typeDefs), getTypeByName);
    expect(resolvers).toMatchSnapshot();
  });
});
