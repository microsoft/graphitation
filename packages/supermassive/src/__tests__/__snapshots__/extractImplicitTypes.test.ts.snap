// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractImplicitTypes benchmark schema extract 1`] = `
Object {
  "Alive": Object {
    "__resolveType": undefined,
    "__types": Array [
      "Person",
      "Species",
    ],
  },
  "Film": Object {},
  "Node": Object {
    "__implementedBy": Array [
      "Film",
      "Vehicle",
      "Person",
      "Starship",
      "Planet",
      "Species",
      "Transport",
    ],
    "__resolveType": undefined,
  },
  "NodeType": "NodeType",
  "Person": Object {},
  "Planet": Object {},
  "Query": Object {},
  "SearchResult": Object {
    "__resolveType": undefined,
    "__types": Array [
      "Person",
      "Starship",
      "Transport",
      "Species",
      "Vehicle",
      "Planet",
      "Film",
    ],
  },
  "Species": Object {},
  "Starship": Object {},
  "Subscription": Object {},
  "Transport": Object {},
  "Vehicle": Object {},
}
`;
